src/nhs_grid.jl

```diff
@@ -471,6 +498,13 @@ end
             if distance2 <= search_radius^2
                 distance = sqrt(distance2)
 
+                # Check if `neighbor_coords` are in the cell `neighbor_cell_`.
+                # For the `SpatialHashingCellList`, this might not be the case
+                # if we have a collision.
+                if check_collision(neighbor_cell_, neighbor_coords, cell_list, neighborhood_search) && cell_collision
```

Comment:
```suggestion
                if cell_collision && check_collision(neighbor_cell_, neighbor_coords, cell_list, neighborhood_search)
```
This is crucial! This is a short-circuit evaluation, which is evaluated from left to right, meaning you always call the function `check_collision`, and then check the flag `cell_collision`.
With my suggestions, you skip the call to `check_collision` for most cells because `cell_collision` is false.

This should also greatly improve performance.